### PR TITLE
Fix -1e4 as attn mask

### LINKED ISSUE #17215: -1e9 constants in T5 implementation
These -1e9 constants are too large for fp16 training

https://github.com/huggingface/transformers/blob/df735d1317994e366ab0edff6c55930e18912b7c/src/transformers/models/t5/modeling_tf_t5.py#L728
https://github.com/huggingface/transformers/blob/df735d1317994e366ab0edff6c55930e18912b7c/src/transformers/models/t5/modeling_tf_t5.py#L746

Maybe they should be made configurable?

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Fix the issues regarding `-1e4` as attention mask.

Fix #17215 #17121 #14859

### CODE REVIEW

File: src/transformers/models/gpt2/modeling_gpt2.py
Comment (ydshieh) [inline src/transformers/models/gpt2/modeling_gpt2.py:202]: `self.masked_bias` is therefore not used anymore.

File: src/transformers/modeling_utils.py
Comment (ydshieh) [inline src/transformers/modeling_utils.py:554]: Is it good to have this function?

File: tests/models/t5/test_modeling_tf_t5.py
Comment (ydshieh) [inline tests/models/t5/test_modeling_tf_t5.py:303]: previously, `create_and_check_t5_decoder_model_past_large_inputs` actually get `token_labels` as `attention_mask`. Don't know why the test passed before.

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:558]:
```suggestion
    return torch.finfo(dtype).min
```
I think returning the value is better here.

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:554]:
```suggestion
def mask_value(dtype):
```

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:648]: `self.device` shouldn't be used here as it breaks model parallelism. If we accept the changes to `mask_value` above, this will be a float and there won't be any need to put it on any device manually.
```suggestion
        encoder_extended_attention_mask = (1.0 - encoder_extended_attention_mask) * mask_value(self.dtype)
```

**Review by sgugger — COMMENTED** [review]
The PyTorch implementation relies on `self.device` which breaks the model parallelism for big model inference, so we should avoid using it (I actually removed lots of instance where we used it recently, and will hunt the other ones in another PR ;-) )

File: src/transformers/modeling_flax_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_flax_utils.py:73]: IMO it's not worth a funtion for a single line. That "over-abstracted" would prefer to have this line always in the modeling forward pass

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:88]: same comment here

File: src/transformers/modeling_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:554]: same comment here

**Comment by patrickvonplaten** [comment]
Generally, this looks good to me. I'd prefer though to not factor out a one-liner into a function (even if we have to add the one-liner 100+ times). It's not good for readability to have to jump to `modeling_utils.py` and the code saved is not worth it for a one-liner IMO.

Also, I'd advocate to make three seperate PRs (one for PT, one for TF, one for Flax). Think it should be both easier to maintain the PRs as well as review them. 

A first test should then be that all slow tests pass. After that it would indeed be nice if we could run some fine-tuning for the most important models (BERT on GLUE, GPT2 on causal LM, T5 on translation maybe). Maybe also not even necessary to verify that everything is correct with a training run if the slow tests all pass

**Comment by ydshieh** [comment]
Hi,

@patrickvonplaten:
  - I removed the new function.
  - I have to modify `FlaxT5Attention` otherwise the PT/Flax T5 equivalence tests will fail.

@sgugger:
  - since there is no more new function `mask_value()`, so no more `device` issue. There is one place I need to use tensor and device though:
  https://github.com/huggingface/transformers/blob/195ef42e0be974e8c019e9d5f03070f65365c721/src/transformers/models/gpt2/modeling_gpt2.py#L202-L205

**Review by patil-suraj — COMMENTED** [review]
It looks good in general. I have pretty much the same comments as Patrick. I would advocate to do some fine-tuning even if the slow tests pass to make sure it doesn't break anything. Especially with models like T5 which have had issues with `attention_mask`.

**Comment by LysandreJik** [comment]
Cool, exciting!

**Comment by ydshieh** [comment]
Hi, @patrickvonplaten @patil-suraj @sgugger @LysandreJik 

This PR is ready for review. 

- Only dealing with PyTorch models: but need to change `FlaxT5` too to make the test pass.
- In general, change to `torch.finfo(correct-dtype).min` instead of `-10000`, `-1e9` etc.
- In particular, changes in `modeling_utils.py`
- Verified the change by training a T5 from scratch as well as finetuning the `t5-small` checkpoint

**Review by sgugger — APPROVED** [review]
Thanks for fixing all of those!

File: src/transformers/models/data2vec/modeling_data2vec_audio.py
Comment (patrickvonplaten) [inline src/transformers/models/data2vec/modeling_data2vec_audio.py:580]: nit
```suggestion
            attention_mask = (1.0 - attention_mask[:, None, None, :].to(dtype=hidden_states.dtype))
            attention_mask = attention_mask * torch.finfo(hidden_states.dtype).min
```

File: src/transformers/models/hubert/modeling_hubert.py
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:666]: (nit) same - maybe nicer to have it in two lines. But feel free to disregard if it's too much work

**Review by patrickvonplaten — APPROVED** [review]
Great work @ydshieh ! Looks good to me

**Comment by ydshieh** [comment]
@sgugger @patrickvonplaten: sorry to call you after the approval, there is something more to discuss after I saw #17437 by @younesbelkada.

- First I need to remove the use of `float("-inf")` as done in [this commit](https://github.com/huggingface/transformers/pull/17306/commits/9fa9d9e8ac4aec9df3c42b3c6e6b178271230247)
  - It's better not to use `inf`: it is hard to track if there is any arithmetic ops that will produce `NaN`.

- Second, there are still some issues when using `torch.finfo(dtype).min`, especially when running in `fp16`, see the code snippet below.

  Basically: 
  - `torch.finfo(torch.float16).min + (-16.0) = -inf`. (or anything < `-16.0`)
  - In some cases, we get all 0s as attention mask:
    - for example, `[pad_token, token_1]` will give an attention mask `[[0, 0], [0, 1]]` (due to causal mask + padding).
    - this mask becomes `[[-65504, -65504], ...]` after using `torch.finfo(dtype).min` (for `fp16`)
    - This might give `[[-inf, -inf], ...]` when combining an `atten_scores` with values < -16.0 
    - Then, we get `NaN` after softmax. 

- `torch`'s `softmax` can't work with fp16 input on CPU
  - RuntimeError: "softmax_lastdim_kernel_impl" not implemented for 'Half' 

### Suggestions

- Cast `fp16` to `fp32` just before `softmax`, so it can run on CPU
- Perform some other processing before `softmax` to avoid `NaN` and non-sense output, especially avoid `[-inf, -inf]` or `[-inf, - large_value]` as input  to `softmax`
- Change attn probability to `[0, 0, .. 0]` if the input is all large negative value `torch.finfo(dtype).min` 

### Code Snippet
```
import torch
from torch import nn

# device = "cpu" --> not working with softmax on float16 (RuntimeError: "softmax_lastdim_kernel_impl" not implemented for 'Half')
device = "cuda"

dtype = torch.float16

mask_value = torch.finfo(dtype).min
attn_mask = torch.tensor([mask_value, mask_value], dtype=dtype)

attn_scores_0 = torch.tensor([-0, -0], dtype=dtype)
attn_scores_1 = torch.tensor([-4, -16], dtype=dtype)
attn_scores_2 = torch.tensor([-18, -16], dtype=dtype)

final_attn_scores_0 = attn_scores_0 + attn_mask  # --> [-65504,  -inf]
final_attn_scores_1 = attn_scores_1 + attn_mask  # --> [-65504,  -inf]
final_attn_scores_2 = attn_scores_2 + attn_mask  # --> [-inf,  -inf]

attn_prob_0 = nn.functional.softmax(final_attn_scores_1.to(device), dim=-1)  # --> [0.5, 0.5], but non-sense!!
attn_prob_1 = nn.functional.softmax(final_attn_scores_1.to(device), dim=-1)  # --> [1, 0], but non-sense!!
attn_prob_2 = nn.functional.softmax(final_attn_scores_2.to(device), dim=-1)  # --> [nan, nan], very bad!

print(final_attn_scores_0)
print(final_attn_scores_1)
print(final_attn_scores_2)
print(attn_prob_0)
print(attn_prob_1)
print(attn_prob_2)
```

**Comment by sgugger** [comment]
Nothing should be done in FP16 on CPU, softmax is not the only operation that is not implemented in Half on CPU.

**Comment by ydshieh** [comment]
> Nothing should be done in FP16 on CPU, softmax is not the only operation that is not implemented in Half on CPU.

OK, thank you! Still want to hear your opinion about other points when you have more time

**Comment by sgugger** [comment]
I'm personally leaning toward implementing an util that preprocesses the attention mask before the softmax as you suggested @ydshieh, but curious to see others opinion.

**Comment by ydshieh** [comment]
I think we can leave the discussion regarding `softmax` and  `attention score processing` to another thread and PR.

After removing `float("-inf")` in this PR, it's less likely to get all `-inf` in a particular sequence. (It's sill likely to happen during training, so better to have a process done after this PR).

So I would prefer to merge the current version once @LysandreJik is happy with this PR.

File: src/transformers/models/albert/modeling_albert.py
Comment (stephenroller) [inline src/transformers/models/albert/modeling_albert.py:731]: imho, the incantation is obtuse enough to make a utility function:

```python
def near_neg_inf(dtype):
    return torch.finfo(self.dtype).min
```

Comment (sgugger) [reply src/transformers/models/albert/modeling_albert.py:731]: I personally prefer to avoid those kinds of one-line functions as it makes it harder to read the code, with the reader having to navigate.

**Comment by ydshieh** [comment]
@michaelbenayoun 

I have tried to keep your change 
```
mask = torch.full((tgt_len, tgt_len), torch.tensor(float("-inf")))
```
with my own, so it becomes
```
mask = torch.full((tgt_len, tgt_len), torch.tensor(torch.finfo(dtype).min))
```

If you encounter any problem after this PR is merged, don't hesitate to ping me.

**Comment by ydshieh** [comment]
@michaelbenayoun I need you help 🙏 


https://github.com/huggingface/transformers/blob/575a8c0a8e2bf6491d7ef0f932fb6caf8a7712b1/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L135

I need to change `torch.tensor(float("-inf")))` to `torch.tensor(torch.finfo(dtype of an input).min))`

but I can't understand what the input it is here. Is it `g` or is it `self`.

From the line below
```
masked_fill(g, output, r_mask, g.op("Constant", value_t=torch.tensor(0, dtype=torch.uint8)))
```
I guess it might be the second argument, so I should use `self`, but it looks strange to me 😢

**Comment by ydshieh** [comment]
> @ydshieh About the first change you mention, I don't think it will break anything, and I will try to apply the same changes when I add support for new model architectures.
> 
> About the symbolic function, you are right I think. Basically `self` is `input` here, maybe we should change the name of the parameter to make things clearer?

Great, thanks for the answer! We can change the name, but not very urgent.

**Review by LysandreJik — APPROVED** [review]
LGTM!

File: src/transformers/modeling_utils.py
Comment (ydshieh) [inline src/transformers/modeling_utils.py:778]: Add a `dtype` argument, so we can use it in `LayoutLMv3Model`

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py
Comment (ydshieh) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:891]: Need to pass the `embedding_output.dtype` in order to make `LayoutLMv3ModelTest.test_multi_gpu_data_parallel_forward` works.

**Review by ydshieh — COMMENTED** [review]
@patrickvonplaten @sgugger Could you take a look of the latest changes below?

- Currently, only `LayoutLMv3Model` uses the new argument `dtyte`.
- In a follow-up PR, we can use this argument for other models
  - it is not straightforward: we need to compute `embedding_output` before we can use it, which is **NOT** the case for `Bert` etc. so far.
  - The change in this PR is quite large already. Better to make a follow-up PR for `dtype`.
- Should do the same thing (dtype argument) for `invert_attention_mask` .

### TODO [**Important**]
We need to address the issue about all `-inf` attention score (in a sequence) and therefore `NaN` for attention probably (after `softmax`), as mentioned in #17433. This happens mostly in `fp16`, in particularly, during training.

File: src/transformers/modeling_utils.py
Comment (ydshieh) [reply src/transformers/modeling_utils.py:778]: The other 2 changes (not shown above)

https://github.com/huggingface/transformers/blob/de8beb96c1d0d323d5d09a85438b47b47a52b0ac/src/transformers/modeling_utils.py#L730

https://github.com/huggingface/transformers/blob/de8beb96c1d0d323d5d09a85438b47b47a52b0ac/src/transformers/modeling_utils.py#L744-L745

**Comment by ydshieh** [comment]
Will merge today (as been approved by 3 core maintainers), if there is no further comment on

https://github.com/huggingface/transformers/pull/17306#pullrequestreview-994918858

**Comment by ydshieh** [comment]
As discussed on Slack, will wait the release of Bloom, and not merge now.

**Pull request merged**
